src/validators.ts

```diff
@@ -0,0 +1,72 @@
+'use strict';
+
+import BN from 'bn.js';
+import { CurrentEpochValidatorInfo, NextEpochValidatorInfo } from './providers/provider';
+
+/** Finds seat price given validators stakes and number of seats.
+ *  Calculation follow the spec: https://nomicon.io/Economics/README.html#validator-selection
+ * @params validators: current or next epoch validators.
+ * @params numSeats: number of seats.
+ */
+export function findSeatPrice(validators: (CurrentEpochValidatorInfo | NextEpochValidatorInfo)[], numSeats: number): BN {
+    const stakes = validators.map(v => new BN(v.stake, 10)).sort((a, b) => a.cmp(b));
+    const num = new BN(numSeats);
+    const stakesSum = stakes.reduce((a, b) => a.add(b));
+    if (stakesSum.lt(num)) {
+        throw 'Stakes are below seats';
+    }
+    // assert stakesSum >= numSeats
+    let left = new BN(1), right = stakesSum.add(new BN(1));
+    while (!left.eq(right.sub(new BN(1)))) {
+        const mid = left.add(right).div(new BN(2));
+        let found = false;
+        let currentSum = new BN(0);
+        for (let i = 0; i < stakes.length; ++i) {
+            currentSum = currentSum.add(stakes[i].div(mid));
+            if (currentSum.gte(num)) {
+                left = mid;
+                found = true;
+                break;
+            }
+        }
+        if (!found) {
+            right = mid;
+        }
+    }
+    return left;
+}
+
+export class ChangedValidatorInfo {
+    public current: CurrentEpochValidatorInfo;
+    public next: NextEpochValidatorInfo;
+
+    public constructor(init?: Partial<ChangedValidatorInfo>) {
+        Object.assign(this, init);
+    }
+}
+
+export class EpochValidatorsDiff {
+    newValidators: NextEpochValidatorInfo[];
+    removedValidators: CurrentEpochValidatorInfo[];
+    changedValidators: ChangedValidatorInfo[];
+
+    public constructor(init?: Partial<EpochValidatorsDiff>) {
+        Object.assign(this, init);
+    }
+}
+
+/** Diff validators between current and next epoch.
+ * Returns additions, subtractions and changes to validator set.
+ * @params 
+ */
+export function diffEpochValidators(currentValidators: CurrentEpochValidatorInfo[], nextValidators: NextEpochValidatorInfo[]): EpochValidatorsDiff {
+    const validatorsMap = new Map<string, CurrentEpochValidatorInfo>();
+    currentValidators.forEach(v => validatorsMap[v.account_id] = v);
```

Comment:
I think `[]` operator on `Map` is just using it as plain JS object. Like you need to use `.put` and `.get` if using it as Map (JS is weird). Not like it matters much for string keys, but then you can just use `{}` instead of `new Map()`.

also you can create `Map` based on key value pairs like this:

```
const validatorsMap = new Map(currentValidators.map(v => [v.account_id, v]));
```